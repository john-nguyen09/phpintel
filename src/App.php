<?php
declare(strict_types=1);

namespace PhpIntel;

use League\Container\Container;
use Microsoft\PhpParser\Parser;

final class App
{
    /**
     * @var Container
     */
    private static $container;

    public static function init()
    {
        if (self::$container === null) {
            self::$container = new Container();
        }

        self::bind();
    }

    public static function make(string $alias)
    {
        return self::$container->get($alias);
    }

    protected static function bind()
    {
        self::$container->share('parser', new Parser);
    }
}