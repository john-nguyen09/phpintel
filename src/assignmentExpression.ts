import * as Parser from "tree-sitter";
import { Variable } from "./variable";
import { TreeAnalyser } from "./treeAnalyser";
import { TypeResolver } from "./typeResolver";

export class AssignmentExpression {
    private _isVariable: boolean = false;
    private _variable: Variable = new Variable();

    constructor(node: Parser.SyntaxNode) {
        let hasEqual = false;

        for (const child of node.children) {
            if (child.type === '=') {
                hasEqual = true;
                continue;
            }

            if (!hasEqual) {
                if (child.type === 'variable_name') {
                    this._variable.name = child.text;
                    continue;
                }
            } else {
                const type = TypeResolver.getNodeType(child);

                if (!type.isEmpty) {
                    this._variable.type.push(type);
                }

                break; // TODO: There can be multiple nodes after the equal sign need to handle this
            }
        }
    }

    get isVariable(): boolean {
        return this._isVariable;
    }

    get variable(): Variable {
        return this._variable;
    }
}