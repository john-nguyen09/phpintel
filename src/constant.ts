import { Location } from "./meta";
import { Type, TypeComposite } from "./typeResolver/type";

export class Constant {
    public name: Type = new Type('');
    public description: string = '';
    public value: string = '';
    public type: TypeComposite = new TypeComposite();
    public location: Location | undefined = undefined;
}