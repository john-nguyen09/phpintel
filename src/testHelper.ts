import { PhpDocument } from "./symbol/phpDocument";
import { pathToUri } from "./util/uri";
import * as fs from "fs";
import { SymbolParser } from "./symbol/symbolParser";
import { Traverser } from "./traverser";
import { TreeNode } from "./util/parseTree";
import { Parser, Phrase, phraseTypeToString, tokenTypeToString } from "php7parser";
import * as path from "path";
import { inspect } from "util";

export function indexFiles(filePaths: string[]): PhpDocument[] {
    let phpDocs: PhpDocument[] = [];
    let treeTraverser = new Traverser();

    for (let filePath of filePaths) {
        const fileUri = pathToUri(filePath);
        const fileContent = fs.readFileSync(filePath).toString();
        const symbolParser = new SymbolParser(new PhpDocument(fileUri, fileContent));
        const parseTree = Parser.parse(fileContent);
        
        treeTraverser.traverse(parseTree, [symbolParser]);

        phpDocs.push(symbolParser.getTree());
    }

    return phpDocs;
}

export function getCaseDir(): string {
    return path.resolve(__dirname, "..", "case");
}

export function getDebugDir(): string {
    return path.resolve(__dirname, '..', 'debug');
}

export function dumpToDebug(name: string, object: any, depth?: number): void {
    fs.writeFile(
        path.resolve(__dirname, '..', 'debug', name),
        inspect(object, {
            depth: depth
        }),
        (err) => {
            console.error(err);
        }
    );
}

export function dumpAstToDebug(name: string, parseTree: Phrase): void {
    fs.writeFile(
        path.resolve(__dirname, '..', 'debug', name),
        JSON.stringify(parseTree, (key, value) => {
            if (key == 'modeStack') {
                return undefined;
            }

            if (key == 'phraseType') {
                return phraseTypeToString(value);
            }

            if (key == 'tokenType') {
                return tokenTypeToString(value);
            }

            return value;
        }, 2),
        (err) => {
            if (err) {
                console.log(err);
            }
        }
    );
}