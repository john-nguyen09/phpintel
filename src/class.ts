import { Location } from "./meta";
import { Interface } from "./interface";
import { Constant } from "./constant";
import { Property } from "./property";
import { Method } from "./method";
import { ClassModifier } from "./modifier";

export class Class {
    public modifier: ClassModifier = ClassModifier.None;

    public name: string = '';
    public location: Location | undefined = undefined;
    public description: string = '';

    public extends: string[] = [];
    public implements: string[] = [];

    public constants: Constant[] = [];
    public properties: Property[] = [];
    public methods: Method[] = [];
}