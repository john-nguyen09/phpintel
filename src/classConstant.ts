import { Constant } from "./constant";
import { VisibilityModifier } from "./modifier";

export class ClassConstant extends Constant {
    public visibility: VisibilityModifier = VisibilityModifier.Public;
    public scope: string | undefined = undefined;

    public extends(constant: Constant) {
        this.name = constant.name;
        this.value = constant.value;
        this.type = constant.type;
        this.description = constant.description;
        this.location = constant.location;
    }
}