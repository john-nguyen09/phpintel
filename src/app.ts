import "reflect-metadata";
import { LevelDatasource } from "./storage/db";
import { LogWriter } from "./service/logWriter";
import { Hasher } from "./service/hasher";
import { Container, interfaces } from "inversify";
import { Indexer } from "./index/indexer";
import { ClassTable } from "./storage/table/class";
import { ClassConstantTable } from "./storage/table/classConstant";
import { ConstantTable } from "./storage/table/constant";
import { FunctionTable } from "./storage/table/function";
import { MethodTable } from "./storage/table/method";
import { PropertyTable } from "./storage/table/property";
import { Traverser } from "./traverser";
import { ReferenceTable } from "./storage/table/reference";
import { PhpDocumentTable } from "./storage/table/phpDoc";
import { IConnection } from "vscode-languageserver";
import { ScopeVarTable } from "./storage/table/scopeVar";

export interface AppOptions {
    storage: string;
}

export class Application {
    protected container: Container = new Container();
    protected options: AppOptions = {
        storage: ''
    };

    constructor(storageLocation: string, connection?: IConnection) {
        this.container.bind<IConnection | undefined>('IConnection').toConstantValue(connection);

        this.initBind();
        this.initStorage(storageLocation);
    }

    public getContainer(): Container {
        return this.container;
    }

    public async clearCache() {
        const db = this.container.get<LevelDatasource>(LevelDatasource).getDb();
        let promises: Promise<void>[] = [];

        return new Promise<void>((resolve, reject) => {
            db.createKeyStream()
                .on('data', (key) => {
                    promises.push(db.del(key));
                })
                .on('end', () => {
                    Promise.all(promises).then(() => { resolve() });
                })
                .on('error', (err: Error) => {
                    reject(err);
                });
        });
    }

    public async shutdown() {
        await this.container.get<LevelDatasource>(LevelDatasource).getDb().close();
    }

    protected initStorage(location: string) {
        this.options.storage = location;

        let datasource = new LevelDatasource(location);
        this.container.bind<LevelDatasource>(LevelDatasource).toConstantValue(datasource);

        // Tables
        this.container.bind<ClassTable>(ClassTable).toSelf().inSingletonScope();
        this.container.bind<ClassConstantTable>(ClassConstantTable).toSelf().inSingletonScope();
        this.container.bind<ConstantTable>(ConstantTable).toSelf().inSingletonScope();
        this.container.bind<FunctionTable>(FunctionTable).toSelf().inSingletonScope();
        this.container.bind<MethodTable>(MethodTable).toSelf().inSingletonScope();
        this.container.bind<PropertyTable>(PropertyTable).toSelf().inSingletonScope();
        this.container.bind<PhpDocumentTable>(PhpDocumentTable).toSelf().inSingletonScope();
        this.container.bind<ReferenceTable>(ReferenceTable).toSelf().inSingletonScope();
        this.container.bind<ScopeVarTable>(ScopeVarTable).toSelf().inSingletonScope();
    }

    protected initBind() {
        this.container.bind<LogWriter>(LogWriter).toSelf().inSingletonScope();
        this.container.bind<Hasher>(Hasher).toSelf().inSingletonScope();
        this.container.bind<Indexer>(Indexer).toSelf().inSingletonScope();
        this.container.bind<Traverser>(Traverser).toSelf().inSingletonScope();
    }
}

export namespace App {
    let app: Application;

    export function init(storageLocation: string, connection?: IConnection) {
        app = new Application(storageLocation, connection);
    }

    export function get<T>(
        identifier: string | symbol | interfaces.Newable<T> | interfaces.Abstract<T>
    ): T {
        return app.getContainer().get(identifier);
    }

    export function set<T>(
        identifier: string,
        value: T
    ): void {
        app.getContainer().bind<T>(identifier).toConstantValue(value);
    }

    export async function clearCache() {
        return await app.clearCache();
    }

    export async function shutdown() {
        return await app.shutdown();
    }
}
