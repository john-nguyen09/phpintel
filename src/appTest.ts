import { Application, IApp } from "./app";
import { IConnection } from "vscode-languageserver";
import { interfaces } from "inversify";
import * as path from "path";

class TestApplication extends Application implements IApp {
    public run() { }
}

export namespace App {
    const app: IApp = new TestApplication();

    app.initStorage(path.join(__dirname, '..', 'debug', 'storage'));
    
    export function run() {
        app.run();
    }

    export function get<T>(
        identifier: string | symbol | interfaces.Newable<T> | interfaces.Abstract<T>
    ): T {
        return app.get(identifier);
    }
}
