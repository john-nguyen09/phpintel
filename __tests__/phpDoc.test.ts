import { indexFiles, getCaseDir, dumpToDebug, dumpAstToDebug } from "../src/testHelper";
import * as path from "path";

describe('phpDoc', () => {
    it('snapshot of phpDocs', () => {
        let phpDocs = indexFiles([
            path.join(getCaseDir(), 'class_constants.php'),
            path.join(getCaseDir(), 'class_methods.php'),
            path.join(getCaseDir(), 'function_declare.php'),
            path.join(getCaseDir(), 'global_symbols.php'),
        ]);

        for (let phpDoc of phpDocs) {
            // dumpAstToDebug(path.basename(phpDoc.uri) + '.ast.json', phpDoc.getTree());
            expect(phpDoc.toObject()).toMatchSnapshot();
        }
    });
});