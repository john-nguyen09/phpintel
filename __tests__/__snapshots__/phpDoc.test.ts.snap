// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`phpDoc snapshot of phpDocs 1`] = `
PhpDocument {
  "classConstants": Array [
    ClassConstant {
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 21,
            "line": 2,
            "offset": 50,
          },
          "start": Position {
            "character": 10,
            "line": 2,
            "offset": 39,
          },
        },
        "relativeUri": "case/class_constants.php",
      },
      "name": TypeName {
        "isArray": false,
        "name": "CONST1",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithConst",
      },
    },
    ClassConstant {
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 19,
            "line": 3,
            "offset": 71,
          },
          "start": Position {
            "character": 10,
            "line": 3,
            "offset": 62,
          },
        },
        "relativeUri": "case/class_constants.php",
      },
      "name": TypeName {
        "isArray": false,
        "name": "PI",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithConst",
      },
    },
    ClassConstant {
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 34,
            "line": 4,
            "offset": 107,
          },
          "start": Position {
            "character": 10,
            "line": 4,
            "offset": 83,
          },
        },
        "relativeUri": "case/class_constants.php",
      },
      "name": TypeName {
        "isArray": false,
        "name": "STATUS_ACTIVE",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithConst",
      },
    },
    ClassConstant {
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 38,
            "line": 5,
            "offset": 147,
          },
          "start": Position {
            "character": 10,
            "line": 5,
            "offset": 119,
          },
        },
        "relativeUri": "case/class_constants.php",
      },
      "name": TypeName {
        "isArray": false,
        "name": "STATUS_INACTIVE",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithConst",
      },
    },
    ClassConstant {
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 26,
            "line": 6,
            "offset": 175,
          },
          "start": Position {
            "character": 10,
            "line": 6,
            "offset": 159,
          },
        },
        "relativeUri": "case/class_constants.php",
      },
      "name": TypeName {
        "isArray": false,
        "name": "IS_ACTIVE",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithConst",
      },
    },
  ],
  "classes": Array [
    Class {
      "extend": null,
      "implements": Array [],
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 1,
            "line": 7,
            "offset": 178,
          },
          "start": Position {
            "character": 0,
            "line": 1,
            "offset": 6,
          },
        },
        "relativeUri": "case/class_constants.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 0,
      },
      "name": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithConst",
      },
      "traits": Array [],
    },
  ],
  "constants": Array [],
  "functions": Array [],
  "importTable": ImportTable {
    "imports": Object {},
  },
  "methods": Array [],
  "properties": Array [],
}
`;

exports[`phpDoc snapshot of phpDocs 2`] = `
PhpDocument {
  "classConstants": Array [],
  "classes": Array [
    Class {
      "extend": null,
      "implements": Array [],
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 1,
            "line": 55,
            "offset": 1074,
          },
          "start": Position {
            "character": 0,
            "line": 2,
            "offset": 7,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 0,
      },
      "name": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
      "traits": Array [],
    },
  ],
  "constants": Array [],
  "functions": Array [],
  "importTable": ImportTable {
    "imports": Object {},
  },
  "methods": Array [
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 27,
            "offset": 607,
          },
          "start": Position {
            "character": 4,
            "line": 25,
            "offset": 560,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": undefined,
      "name": TypeName {
        "isArray": false,
        "name": "__construct",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 31,
            "offset": 667,
          },
          "start": Position {
            "character": 4,
            "line": 29,
            "offset": 613,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "method1",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 38,
            "offset": 773,
          },
          "start": Position {
            "character": 4,
            "line": 33,
            "offset": 673,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "method2",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 42,
            "offset": 830,
          },
          "start": Position {
            "character": 4,
            "line": 40,
            "offset": 779,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "method3",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 46,
            "offset": 890,
          },
          "start": Position {
            "character": 4,
            "line": 44,
            "offset": 836,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "method4",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 50,
            "offset": 953,
          },
          "start": Position {
            "character": 4,
            "line": 48,
            "offset": 896,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 2,
      },
      "name": TypeName {
        "isArray": false,
        "name": "protectedMethod1",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 54,
            "offset": 1072,
          },
          "start": Position {
            "character": 4,
            "line": 52,
            "offset": 959,
          },
        },
        "relativeUri": "case/class_methods.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 4,
      },
      "name": TypeName {
        "isArray": false,
        "name": "privateMethod1",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\ClassWithMethod",
      },
    },
  ],
  "properties": Array [
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": "$property1",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": "$property2",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 2,
      },
      "name": "$protected1",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 2,
      },
      "name": "$protected2",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 2,
      },
      "name": "$protected3",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 2,
      },
      "name": "$protectedProperty1",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 4,
      },
      "name": "$privateProperty1",
    },
  ],
}
`;

exports[`phpDoc snapshot of phpDocs 3`] = `
PhpDocument {
  "classConstants": Array [],
  "classes": Array [],
  "constants": Array [],
  "functions": Array [
    Function {
      "description": "",
      "name": TypeName {
        "isArray": false,
        "name": "\\\\test_function1",
      },
      "parameters": Array [
        Parameter {
          "name": "$stringParam",
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "isArray": false,
                "name": "string",
              },
            ],
          },
          "value": "",
        },
        Parameter {
          "name": "$intParam",
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "isArray": false,
                "name": "int",
              },
            ],
          },
          "value": "",
        },
        Parameter {
          "name": "$boolParam",
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "isArray": false,
                "name": "boolean",
              },
            ],
          },
          "value": "",
        },
        Parameter {
          "name": "$floatParam",
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "isArray": false,
                "name": "float",
              },
            ],
          },
          "value": "",
        },
      ],
      "scopeVar": Scope {
        "variables": Object {
          "$boolParam": Variable {
            "expression": Expression {},
            "name": "$boolParam",
            "type": TypeComposite {
              "_types": Array [
                TypeName {
                  "isArray": false,
                  "name": "boolean",
                },
              ],
            },
          },
          "$floatParam": Variable {
            "expression": Expression {},
            "name": "$floatParam",
            "type": TypeComposite {
              "_types": Array [
                TypeName {
                  "isArray": false,
                  "name": "float",
                },
              ],
            },
          },
          "$intParam": Variable {
            "expression": Expression {},
            "name": "$intParam",
            "type": TypeComposite {
              "_types": Array [
                TypeName {
                  "isArray": false,
                  "name": "int",
                },
              ],
            },
          },
          "$stringParam": Variable {
            "expression": Expression {},
            "name": "$stringParam",
            "type": TypeComposite {
              "_types": Array [
                TypeName {
                  "isArray": false,
                  "name": "string",
                },
              ],
            },
          },
          "$test": SimpleVariable {
            "expression": Expression {
              "currentSymbol": TokenSymbol {
                "text": "1",
              },
            },
            "name": "$test",
            "type": TypeComposite {
              "_types": Array [
                TypeName {
                  "isArray": false,
                  "name": "int",
                },
              ],
            },
          },
        },
      },
      "types": Array [
        TypeName {
          "isArray": false,
          "name": "boolean",
        },
        TypeName {
          "isArray": false,
          "name": "string",
        },
        TypeName {
          "isArray": false,
          "name": "int",
        },
        TypeName {
          "isArray": false,
          "name": "float",
        },
        TypeName {
          "isArray": false,
          "name": "\\\\TestClass",
        },
      ],
    },
    Function {
      "description": "Test function 2",
      "name": TypeName {
        "isArray": false,
        "name": "\\\\test_function2",
      },
      "parameters": Array [
        Parameter {
          "name": "$param1",
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "isArray": false,
                "name": "\\\\array|int",
              },
            ],
          },
          "value": "",
        },
        Parameter {
          "name": "$param2",
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "isArray": true,
                "name": "string",
              },
            ],
          },
          "value": "",
        },
      ],
      "scopeVar": Scope {
        "variables": Object {
          "$param1": Variable {
            "expression": Expression {},
            "name": "$param1",
            "type": TypeComposite {
              "_types": Array [
                TypeName {
                  "isArray": false,
                  "name": "\\\\array|int",
                },
              ],
            },
          },
          "$param2": Variable {
            "expression": Expression {},
            "name": "$param2",
            "type": TypeComposite {
              "_types": Array [
                TypeName {
                  "isArray": true,
                  "name": "string",
                },
              ],
            },
          },
        },
      },
      "types": Array [],
    },
  ],
  "importTable": ImportTable {
    "imports": Object {},
  },
  "methods": Array [],
  "properties": Array [],
}
`;

exports[`phpDoc snapshot of phpDocs 4`] = `
PhpDocument {
  "classConstants": Array [
    ClassConstant {
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 32,
            "line": 38,
            "offset": 1084,
          },
          "start": Position {
            "character": 10,
            "line": 38,
            "offset": 1062,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "name": TypeName {
        "isArray": false,
        "name": "TEST_CLASS_CONST",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\TestClass",
      },
    },
    ClassConstant {
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 32,
            "line": 100,
            "offset": 2128,
          },
          "start": Position {
            "character": 10,
            "line": 100,
            "offset": 2106,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "name": TypeName {
        "isArray": false,
        "name": "TEST_CLASS_CONST",
      },
      "scope": null,
    },
  ],
  "classes": Array [
    Class {
      "extend": null,
      "implements": Array [
        TypeName {
          "isArray": false,
          "name": "TestInterface",
        },
        TypeName {
          "isArray": false,
          "name": "TestInterface2",
        },
      ],
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 20,
            "offset": 409,
          },
          "start": Position {
            "character": 0,
            "line": 17,
            "offset": 343,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 0,
      },
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TestClass2",
      },
      "traits": Array [],
    },
    Class {
      "extend": null,
      "implements": Array [
        TypeName {
          "isArray": false,
          "name": "TestInterface",
        },
      ],
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 72,
            "offset": 1875,
          },
          "start": Position {
            "character": 0,
            "line": 31,
            "offset": 888,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 0,
      },
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TestClass",
      },
      "traits": Array [],
    },
    Class {
      "extend": TypeName {
        "isArray": false,
        "name": "TestClass",
      },
      "implements": Array [],
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 115,
            "offset": 2385,
          },
          "start": Position {
            "character": 0,
            "line": 115,
            "offset": 2348,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 0,
      },
      "name": TypeName {
        "isArray": false,
        "name": "\\\\ChildClass",
      },
      "traits": Array [],
    },
    Class {
      "extend": null,
      "implements": Array [],
      "location": Location {
        "range": Range {
          "end": Position {
            "character": 1,
            "line": 137,
            "offset": 2883,
          },
          "start": Position {
            "character": 0,
            "line": 130,
            "offset": 2778,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 0,
      },
      "name": TypeName {
        "isArray": false,
        "name": "\\\\UnusedClass",
      },
      "traits": Array [],
    },
  ],
  "constants": Array [
    Constant {
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TEST_CONST",
      },
      "type": TypeName {
        "isArray": false,
        "name": "int",
      },
      "value": "123",
    },
    Constant {
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TEST_CONST1",
      },
      "type": TypeName {
        "isArray": false,
        "name": "int",
      },
      "value": "456",
    },
    Constant {
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TEST_CONST2",
      },
      "type": TypeName {
        "isArray": false,
        "name": "int",
      },
      "value": "789",
    },
    Constant {
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TEST_CONST_STRING",
      },
      "type": TypeName {
        "isArray": false,
        "name": "string",
      },
      "value": "'Simple string'",
    },
    Constant {
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TEST_CONST_STRING2",
      },
      "type": TypeName {
        "isArray": false,
        "name": "string",
      },
      "value": "'Not so simple string'",
    },
    Constant {
      "name": TypeName {
        "isArray": false,
        "name": "\\\\TEST_CONST_STRING3",
      },
      "type": TypeName {
        "isArray": false,
        "name": "string",
      },
      "value": "'Quite complex ' . TEST_CONST_STRING",
    },
  ],
  "functions": Array [
    Function {
      "description": "Officia aliquip adipisicing et nulla et laboris dolore labore.
",
      "name": TypeName {
        "isArray": false,
        "name": "\\\\test_function",
      },
      "parameters": Array [],
      "scopeVar": Scope {
        "variables": Object {},
      },
      "types": Array [],
    },
  ],
  "importTable": ImportTable {
    "imports": Object {},
  },
  "methods": Array [
    Method {
      "description": "Do magna consequat veniam minim proident eiusmod incididunt aute proident.",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 60,
            "offset": 1547,
          },
          "start": Position {
            "character": 4,
            "line": 57,
            "offset": 1457,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "staticTestMethod",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\TestClass",
      },
    },
    Method {
      "description": "Non culpa nostrud mollit esse sunt laboris in irure ullamco cupidatat amet.
",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 71,
            "offset": 1873,
          },
          "start": Position {
            "character": 4,
            "line": 68,
            "offset": 1758,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "testMethod",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\TestClass",
      },
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 107,
            "offset": 2254,
          },
          "start": Position {
            "character": 4,
            "line": 104,
            "offset": 2200,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "staticTestMethod",
      },
      "scope": null,
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 112,
            "offset": 2343,
          },
          "start": Position {
            "character": 4,
            "line": 109,
            "offset": 2260,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "testMethod",
      },
      "scope": null,
    },
    Method {
      "description": "",
      "location": Location {
        "range": Range {
          "end": Position {
            "character": -1,
            "line": 136,
            "offset": 2881,
          },
          "start": Position {
            "character": 4,
            "line": 134,
            "offset": 2831,
          },
        },
        "relativeUri": "case/global_symbols.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": TypeName {
        "isArray": false,
        "name": "unusedMethod",
      },
      "scope": TypeName {
        "isArray": false,
        "name": "\\\\UnusedClass",
      },
    },
  ],
  "properties": Array [
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": "$staticTestProperty",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": "$testProperty",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": "$staticTestProperty",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": "$testProperty",
    },
    Property {
      "description": "",
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 1,
      },
      "name": "$unusedProperty",
    },
  ],
}
`;
